File: ProyectoCivitasJava/src/civitas/JugadorEspeculador.java
/**
 * Proyecto Civitas
 * Sergio Hervás Cobo
 * Javier Suárez Lloréns
 * Programación y Diseño Orientado a Objetos
 * 2º B3 - Grado en Ingeniería Informática
 * E.T.S.I. Informática y de Telecomunicación
 * Universidad de Granada
 */
package civitas;

/**
 *
 * @author sergio
 */
public class JugadorEspeculador extends Jugador {
    
    private static final int factorEspeculador = 2;
    private int fianza;
    
    
    public JugadorEspeculador(Jugador otro, int fianza){
        super(otro);
        this.fianza = fianza;
        for (int k = 0; k < otro.getPropiedades().size(); k++){
            otro.getPropiedades().get(k).actualizaPropietarioPorConversion(this);
        }
    }
    
    
    @Override
    protected boolean debeSerEncarcelado(){
        boolean debeserlo = super.debeSerEncarcelado();
        if(debeserlo && puedoGastar(fianza)){
            debeserlo = false;
            paga(fianza);
            Diario.getInstance().ocurreEvento(getNombre() + " paga la fianza y se libra de la cárcel");
        }
        return debeserlo;
    }
   
    
    @Override
    boolean pagaImpuesto(float cantidad){
        boolean comprobacion = false;
        if(!isEncarcelado()){
            comprobacion = paga(cantidad/2);
        }
        return comprobacion;
    }
    
    @Override
    protected boolean puedoEdificarCasa(TituloPropiedad propiedad){
        boolean puedoedificarcasa = false;
        float precio = propiedad.getPrecioEdificar();
        
        if (super.puedoGastar(precio) && propiedad.getNumCasas() < this.getCasasMax()){
            puedoedificarcasa = true;
        }
        return puedoedificarcasa;
    }
    
    @Override
    protected boolean puedoEdificarHotel(TituloPropiedad propiedad){
        boolean puedoedificarhotel = false;
        float precio = propiedad.getPrecioEdificar();
        if ( puedoGastar(precio) && propiedad.getNumHoteles() < this.getHotelesMax() && propiedad.getNumCasas() >= this.getCasasPorHotel()){
            puedoedificarhotel = true;
        }
        return puedoedificarhotel;
    }        
    
    @Override
    protected int getHotelesMax(){
        return super.getHotelesMax()*factorEspeculador;
    }

    @Override
    protected int getCasasMax(){
        return super.getCasasMax()*factorEspeculador;
    }
    
    @Override
    public String toString(){
      String salida = "Jugador Especulador: " + super.toString();
      return salida;
    } 
}

File: ProyectoCivitasRuby/lib/modelo/jugador_especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "jugador"


module Civitas
  class JugadorEspeculador < Jugador
#    public_class_method :new 

    @@factor_especulador = 2    
    
    def initialize(fianza)     
      @fianza = fianza
    end
    
    def self.nuevo_especulador(jugador, fianza)
        espec = new(fianza)
        espec.copia(jugador)
            
      for propiedad in espec.propiedades
        propiedad.actualiza_propietario_por_conversion(self)
      end
      
      espec
    end
    
    #@Override
    def paga_impuesto(cantidad)
      super(cantidad/2)
    end
    
    
    protected
    #@Override
    def debe_ser_encarcelado
      debeserlo = super
      if(debeserlo && puedo_gastar(@fianza))
        debeserlo = false
        paga(@fianza)
        Diario.instance.ocurre_evento("Fianza pagada. El especulador se libra de la carcel.")
      end
      debeserlo
    end
    
    
    #@Override
    def puedo_edificar_casa(propiedad)
      puedo = false
      precio = propiedad.precioedificar()
      if(puedo_gastar(precio) && propiedad.numcasas < casasmax)
        puedo = true
      end
      puedo
    end
    
    
    #@Override
    def puedo_edificar_hotel(propiedad)
      puedo = false
      precio = propiedad.precioedificar()
      if(puedo_gastar(precio) && propiedad.numhoteles < hotelesmax && propiedad.numcasas >= @@casasporhotel)
        puedo = true
      end
      puedo
    end
    
    #@Override
    def hotelesmax
      Jugador.hotelesmax*@@factor_especulador
    end
    
    
    #@Override
    def casasmax
      Jugador.casasmax*@@factor_especulador
    end
    
    
    #@Override
    public
    def to_s
      salida = "Jugador Especulador #{super.to_s}"
    end

    
  end
end